PR: https://github.com/odoo/odoo/pull/230742

From: ff9592bbf64c9a4ce151a897058275c39b2e6096
From: Amr Yasser (amry)
Date: 2026-01-08 08:58:58

Breaking data model changes scores: del:4.4 + add:0, change matches:
-    time_off_type_id = fields.Many2one('hr.leave.type', string="Time Off Type",
-    accrual_plan_id = fields.Many2one('hr.leave.accrual.plan',
+    accrual_plan_id = fields.Many2one('hr.leave.accrual.plan', compute="_compute_accrual_plan_id",
-    accruals_ids = fields.One2many('hr.leave.accrual.plan', 'time_off_type_id')
-    accrual_count = fields.Float(compute="_compute_accrual_count", string="Accruals count")

Total Changes: 74

[REM] hr_holidays: Drop time_off_type_id from accruals

Previously, Accrual Plans could be linked to time off types. This has been removed since Odoo 16, so this commit aims to remove unused fields, actions, and functions, and inaccessible buttons

Task: 4596117
Part-of: odoo/odoo#230742
Related: odoo/enterprise#96812
Related: odoo/upgrade#8604
Signed-off-by: Bertrand Dossogne (bedo) <bedo@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays/models/hr_leave_accrual_plan.py
+++ b/addons/hr_holidays/models/hr_leave_accrual_plan.py
@@ -13,16 +13,11 @@ class HrLeaveAccrualPlan(models.Model):
 
     active = fields.Boolean(default=True)
     name = fields.Char('Name', required=True)
-    time_off_type_id = fields.Many2one('hr.leave.type', string="Time Off Type",
-        check_company=True, index='btree_not_null',
-        help="""Specify if this accrual plan can only be used with this Time Off Type.
-                Leave empty if this accrual plan can be used with any Time Off Type.""")
     employees_count = fields.Integer("Employees", compute='_compute_employee_count')
     level_ids = fields.One2many('hr.leave.accrual.level', 'accrual_plan_id', copy=True, string="Milestones")
     allocation_ids = fields.One2many('hr.leave.allocation', 'accrual_plan_id',
         export_string_translation=False)
-    company_id = fields.Many2one('res.company', string='Company', domain=lambda self: [('id', 'in', self.env.companies.ids)],
-        compute="_compute_company_id", store="True", readonly=False)
+    company_id = fields.Many2one('res.company', string='Company', domain=lambda self: [('id', 'in', self.env.companies.ids)])
     transition_mode = fields.Selection([
         ('immediately', 'Immediately'),
         ('end_of_accrual', "After this accrual's period")],
@@ -92,14 +87,6 @@ class HrLeaveAccrualPlan(models.Model):
         for plan in self:
             plan.employees_count = allocations_dict.get(plan.id, 0)
 
-    @api.depends('time_off_type_id.company_id')
-    def _compute_company_id(self):
-        for accrual_plan in self:
-            if accrual_plan.time_off_type_id:
-                accrual_plan.company_id = accrual_plan.time_off_type_id.company_id
-            else:
-                accrual_plan.company_id = self.env.company
-
     @api.depends("accrued_gain_time")
     def _compute_is_based_on_worked_time(self):
         for plan in self:
@@ -136,7 +123,7 @@ class HrLeaveAccrualPlan(models.Model):
                 new=True,
                 default_can_be_carryover=self.can_be_carryover,
                 default_accrued_gain_time=self.accrued_gain_time,
-                default_can_modify_value_type=not self.time_off_type_id and not self.level_ids,
+                default_can_modify_value_type=not self.level_ids,
                 default_added_value_type=self.added_value_type,
             ),
         }

--- a/addons/hr_holidays/models/hr_leave_accrual_plan_level.py
+++ b/addons/hr_holidays/models/hr_leave_accrual_plan_level.py
@@ -194,22 +194,20 @@ class HrLeaveAccrualLevel(models.Model):
         for level in self:
             level.sequence = level.start_count * start_type_multipliers[level.start_type]
 
-    @api.depends('accrual_plan_id', 'accrual_plan_id.level_ids', 'accrual_plan_id.time_off_type_id')
+    @api.depends('accrual_plan_id', 'accrual_plan_id.level_ids')
     def _compute_can_modify_value_type(self):
         for level in self:
-            level.can_modify_value_type = not level.accrual_plan_id.time_off_type_id and level.accrual_plan_id.level_ids and level.accrual_plan_id.level_ids[0] == level
+            level.can_modify_value_type = level.accrual_plan_id.level_ids and level.accrual_plan_id.level_ids[0] == level
 
     def _inverse_added_value_type(self):
         for level in self:
             if level.accrual_plan_id.level_ids[0] == level:
                 level.accrual_plan_id.added_value_type = level.added_value_type
 
-    @api.depends('accrual_plan_id', 'accrual_plan_id.level_ids', 'accrual_plan_id.added_value_type', 'accrual_plan_id.time_off_type_id')
+    @api.depends('accrual_plan_id', 'accrual_plan_id.level_ids', 'accrual_plan_id.added_value_type')
     def _compute_added_value_type(self):
         for level in self:
-            if level.accrual_plan_id.time_off_type_id:
-                level.added_value_type = "day" if level.accrual_plan_id.time_off_type_id.request_unit in ["day", "half_day"] else "hour"
-            elif level.accrual_plan_id.level_ids and level.accrual_plan_id.level_ids[0] != level:
+            if level.accrual_plan_id.level_ids and level.accrual_plan_id.level_ids[0] != level:
                 level.added_value_type = level.accrual_plan_id.level_ids[0].added_value_type
             elif not level.added_value_type:
                 level.added_value_type = "day"  # default value

--- a/addons/hr_holidays/models/hr_leave_allocation.py
+++ b/addons/hr_holidays/models/hr_leave_allocation.py
@@ -121,9 +121,8 @@ class HrLeaveAllocation(models.Model):
         ('accrual', 'Accrual Allocation')
     ], string="Allocation Type", default="regular", required=True, readonly=True)
     is_officer = fields.Boolean(compute='_compute_is_officer')
-    accrual_plan_id = fields.Many2one('hr.leave.accrual.plan',
-        compute="_compute_accrual_plan_id", inverse="_inverse_accrual_plan_id", store=True, index='btree_not_null', readonly=False, tracking=True,
-        domain="['|', ('time_off_type_id', '=', False), ('time_off_type_id', '=', holiday_status_id)]")
+    accrual_plan_id = fields.Many2one('hr.leave.accrual.plan', compute="_compute_accrual_plan_id",
+    inverse="_inverse_accrual_plan_id", store=True, index='btree_not_null', readonly=False, tracking=True)
     max_leaves = fields.Float(compute='_compute_leaves')
     leaves_taken = fields.Float(compute='_compute_leaves', string='Time off Taken')
     virtual_remaining_leaves = fields.Float(compute='_compute_leaves', string='Available Time Off')
@@ -264,12 +263,9 @@ class HrLeaveAllocation(models.Model):
         default_holiday_status_id = None
         for allocation in self:
             if not allocation.holiday_status_id:
-                if allocation.accrual_plan_id:
-                    allocation.holiday_status_id = allocation.accrual_plan_id.time_off_type_id
-                else:
-                    if not default_holiday_status_id:  # fetch when we need it
-                        default_holiday_status_id = self._default_holiday_status_id()
-                    allocation.holiday_status_id = default_holiday_status_id
+                if not default_holiday_status_id:  # fetch when we need it
+                    default_holiday_status_id = self._default_holiday_status_id()
+                allocation.holiday_status_id = default_holiday_status_id
 
     @api.depends('holiday_status_id', 'number_of_hours_display', 'number_of_days_display', 'type_request_unit', 'employee_id')
     def _compute_number_of_days(self):
@@ -280,21 +276,11 @@ class HrLeaveAllocation(models.Model):
             elif allocation_unit == 'hour' and allocation.employee_id:
                 allocation.number_of_days = allocation.number_of_hours_display / allocation.employee_id._get_hours_per_day(allocation.date_from)
 
-    @api.depends('holiday_status_id', 'allocation_type')
+    @api.depends('allocation_type')
     def _compute_accrual_plan_id(self):
-        accrual_allocations = self.filtered(lambda alloc: alloc.allocation_type == 'accrual' and not alloc.accrual_plan_id and alloc.holiday_status_id)
-        accruals_read_group = self.env['hr.leave.accrual.plan']._read_group(
-            [('time_off_type_id', 'in', accrual_allocations.holiday_status_id.ids)],
-            ['time_off_type_id'],
-            ['id:array_agg'],
-        )
-        accruals_dict = {time_off_type.id: ids for time_off_type, ids in accruals_read_group}
         for allocation in self:
-            if (allocation.allocation_type == 'regular' and allocation.accrual_plan_id) or allocation.accrual_plan_id.time_off_type_id.id not in (False, allocation.holiday_status_id.id):
+            if (allocation.allocation_type == 'regular' and allocation.accrual_plan_id):
                 allocation.accrual_plan_id = False
-            if allocation.allocation_type == 'accrual' and not allocation.accrual_plan_id:
-                if allocation.holiday_status_id:
-                    allocation.accrual_plan_id = accruals_dict.get(allocation.holiday_status_id.id, [False])[0]
 
     def _inverse_accrual_plan_id(self):
         for allocation in self:

--- a/addons/hr_holidays/models/hr_leave_type.py
+++ b/addons/hr_holidays/models/hr_leave_type.py
@@ -117,8 +117,6 @@ class HrLeaveType(models.Model):
         help="If checked, users can request another leave on top of the ones of this type.")
     elligible_for_accrual_rate = fields.Boolean(string='Eligible for Accrual Rate', compute="_compute_eligible_for_accrual_rate", store=True, readonly=False,
         help="If checked, this time off type will be taken into account for accruals computation.")
-    accruals_ids = fields.One2many('hr.leave.accrual.plan', 'time_off_type_id')
-    accrual_count = fields.Float(compute="_compute_accrual_count", string="Accruals count")
     # negative time off
     allows_negative = fields.Boolean(string='Allow Negative Cap',
         help="If checked, users request can exceed the allocated days and balance can go in negative.")
@@ -335,12 +333,6 @@ class HrLeaveType(models.Model):
         for allocation in self:
             allocation.group_days_leave = grouped_dict.get(allocation.id, 0)
 
-    def _compute_accrual_count(self):
-        accrual_allocations = self.env['hr.leave.accrual.plan']._read_group([('time_off_type_id', 'in', self.ids)], ['time_off_type_id'], ['__count'])
-        mapped_data = {time_off_type.id: count for time_off_type, count in accrual_allocations}
-        for leave_type in self:
-            leave_type.accrual_count = mapped_data.get(leave_type.id, 0)
-
     def _compute_is_used(self):
         leaves_count = self._leaves_count_by_leave_type_id()
         allocations_count = self._allocations_count_by_leave_type_id()
@@ -447,17 +439,6 @@ class HrLeaveType(models.Model):
         }
         return action
 
-    def action_see_accrual_plans(self):
-        self.ensure_one()
-        action = self.env["ir.actions.actions"]._for_xml_id("hr_holidays.open_view_accrual_plans")
-        action['domain'] = [
-            ('time_off_type_id', '=', self.id),
-        ]
-        action['context'] = {
-            'default_time_off_type_id': self.id,
-        }
-        return action
-
     # ------------------------------------------------------------
     # Leave - Allocation link methods
     # ------------------------------------------------------------
